Refactor the abstract interpretation code

martin-cs opened the pull request:
Hi all,
        This is quite a substantial patch set and has been a long time in the making.  It radically restructures the abstract interpreter with the goal of making it easier to implement various advanced features such as merging VSD, context dependent analysis, loop bounds analysis, handling function pointers during AI, etc. and also deprecating the other two abstract interpretation frameworks.

  I have tried my best to make all of the changes backwards compatible, so the results given by existing AI users should not change at all.  The only things this should break are out of tree classes that inherit from ait<domainT> and use the interfaces in an unusual way.  As such there aren't test cases because there should be no user / externally visible effects.  I will start a separate draft PR that adds new features and functionality using this patch to demonstrate some of the new things we can do.

  I realise this is a lot to review but each patch should be self contained and documented enough that they can be reviewed one at a time.  I'm also happy to discuss / talk through the changes with anyone interested.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Refactoring allows on-the-fly creation of domains as needed

This can be a big memory saving for larger programs, especially if
only part of the program is being analysed.  Note that dependency
graph requires the old behaviour preserving as it does some complex
initialisation of domains.  Until we have better initialisation of
domains this will have to remain.

Commit:
Remove the find_state internal interface from ai_baset

find_state is not currently used nor is it likely to be.
During computation, *this is non-const so get_state can be used.
Output methods now use the public interface of
abstract_state_before.
The only real space for it to be used is in ait::operator[] const
which can access the map directly so doesn't need to have an
additional interface in ai_baset.

Commit:
Move concurrency specific functionality into concurrency_aware_ait

Although the concurrent fix-point and merge_shared could be in
ai_baset, there is relatively little point.  Both are specific to
the way that concurrency_aware_ait works.  The analysis itself is
only sound if some relatively particular and quite limiting
conditions hold on the domain.  If we want to improve the way we
do abstract intepretation of concurrent programs I doubt we will
want to build on top of this.

Commit:
Clang-format the concurrent fix-point

Commit:
Change the const access to domains from unique_ptr to shared_ptr

A slight increase in overhead in return for an opportunity to
reduce the number of copies when refactoring the domain storage.

Commit:
Remove mutable access to the domains in an ait instance

External, mutable access to individual domains is not really used
by the current applications.  Removing it clarifies ownership of
the domains by the interpreter, makes it harder to get unsound or
inconsistent results from the analyses and makes it easier to
break the "one domain per location" link.

If you need this functionality, there is nothing to stop you
inheriting from ait and re-adding it, although there is likely a
better design of UI you could use (including things like re-running
the fixpoint algorithm after modifications).

Commit:
Store shared pointers to domains rather than domains themselves

This costs one extra shared_ptr per location but it removes the
need to copy domains on access and will enable for exciting new
features in the future.

Commit:
Create a domain factory object so domains can have constructors

The lack of a way of passing configuration options and the like
to domains from outside of the abstract interpreter has been a
long-standing pain.  This moves domain construction into a factory
so that users of the abstract interpreter can just inherit from
ai_domain_factory_baset or ai_domain_factoryt<domainT> to have
an object that carries the extra constructor arguments and then
uses them when needed.

Commit:
Refactor the reaching definition analysis to use the domain factory

This removes an awful hack where get_state is used to set a
construction parameter on access rather than it being, you know,
in the constructor.

Commit:
Remove domain initialisation in favour of make_top/bottom/entry

This assignment is done by make_top/bottom/entry and done more
correctly than this, making this unnecessary.

Commit:
Remove duplicated initialisation pass

The base initialisation pass will run for all functions, so there
is no need to overide this.  It is also idempotent so running it
twice is unnecessary.

Commit:
Refactor invariant_propagation to use a domain factory

This allows a number of pointers to be replaced with reference
among other clean ups.

Commit:
Refactor the dependency graph analysis to use the domain factory

This removes another instance of get_state being used to construct the
domain elements.

Commit:
Remove duplicate initialisation during instrumentation

For on-the-fly domain creation to work get_state shouldn't have any
observable side-effects.  At the moment this is not the case because
the operator[] access will throw an exception if the location is not
initialised.  Thus we can't remove both of the calls to get_state
but we can definitely remove these duplicate ones.

Commit:
Move the domain-specific call of merge into the factory

This is part of a larger effort to move all of the domain-specific
code out of ait so it can just manipulate the domain interface and
doesn't need to know the actual type of the domains.

Commit:
Move the domain factory from ait into the base class

The rest of the architecture has been smoothed enough to do this.
This is part of a goal of removing functionality from ait so that
it can be modularised and extended in other ways.

Commit:
Delegate the storage of domains to an object owned by ai_baset

By delegating the storage of domains to a separate object (in a
similar way to the creation and other domain specific things in
ai_domain_factory) we gain the ability to play with and modify the
number and way that domains are stored without needing to inherit
from ai_baset and re-implement things.  This creates new
possibilities such as pruning domains during analysis, have one
domain per function or even just one over-all (needed to obsolete
some of the older AI frameworks).  It also means that inheritance
from ai_baset or ait can be used to alter how the algorithm works
without being tied to a particular storage model.

Commit:
Refactor and tidy how function calls are handled in AI

The current ai_baset has a "skip if not available, otherwise recurse"
policy which cannot be altered or removed.  Recursion causes problems
for domains with large height as they can crash the analyser.  It also
mixes the (interprocedural) state of the /analysis/ with the recursive
state of the /analyser/.  This forces an undesireable design choice --
it means that anything interprocedural must be handled purely in the
domain.  Doing so has several limitations and so this patch allows
inheritance to change how interprocedural analysis is done.

Commit:
Add a typedef for "a pointer to a domain" to simplify interfaces

These are used because it allows const analysers / ai_storage
objects to generate new (bottom) domains, so on-the-fly domain
creation is properly supported.

Commit:
Replace locations with history abstractions in the abstract interpreter

This is a fundamental change in how the abstract intepreter works.
Previously it's working unit was a location.  Transforms were between
locations and the most precise unit for domain storage was per location.
History objects are abstractions that represent one or more control
flow path histories.  A single location is the most coarse grains of
these but they can include calling contexts, paths taken, loop iterations
thread changes, etc.  This change should be backwards compatible but
also allow more sophisticated histories, giving loop unwinding, context
sensitivity, path sensitivity, thread awareness and many others.

Commit:
One domain per history storage in abstract interpretation

This allows more sophisticated history abstractions that have
more than on distinct history per location (for example loop
unwinding and calling context).

Commit:
Update the documentation in the key abstract interpreter classes

This reflects the changes in the structure and best practice that
result from this series of patches.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8017845487d0410bb1c23b157ce560cf1f899769).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123863950

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -543,24 +513,107 @@ class concurrency_aware_ait:public ait<domainT>
   }
 
-  bool merge_shared(
+  virtual bool merge_shared(
```
Why is this virtual? Is this class supposed to be inherited from?

martin-cs replied:
Not really.  I mean, ideally, `concurrency_aware_ait` would be deprecated and replaced with a solution that would work with other domains.  With the new storage it should be pretty easy to merge the relevant locations and VSD's three-way-merge would allow concurrent AI to be done properly.  But, right now, I don't have the time to refactor all of that.

I made it virtual so that it is consistent with the merge method in ai_baset.

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -100,7 +100,8 @@ class ai_baset
   ///   the method should be const and there are some non-trivial cases
   ///   including merging abstract states, etc.
-  virtual std::unique_ptr<statet> abstract_state_before(locationt l) const = 0;
+  virtual std::shared_ptr<const statet>
```
This one is a bit hard to follow; While the changes themselves are obvious enough it's not easy to understand what's meant by 'refactoring the domain storage' here (maybe that'll make more sense later on in the PR?)

martin-cs replied:
It's referring to the refactoring of how domains are stored that happens 11 commits later.  Although it looks like it from the history, this PR wasn't really written in commit order, or, where it was, it was written with a lot of fore-knowledge of what would happen.

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -395,16 +395,6 @@ class ait:public ai_baset
   typedef goto_programt::const_targett locationt;
 
-  /// Find the analysis result for a given location.
-  domainT &operator[](locationt l)
```
Typo in 46c10cebf8d9589bfa75c787b01ea6faf9db274a: "Clarrifies -> Clarifies"

martin-cs replied:
Fixed.

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -431,16 +431,25 @@ class ait:public ai_baset
+    typename state_mapt::const_iterator it = state_map.find(l);
+    if(it == state_map.end())
+    {
+      auto p = state_map.insert(std::make_pair(l, std::make_shared<domainT>()));
```
You could also use `emplace` to skip the `make_pair`

martin-cs replied:
Changed.

hannes-steffenhagen-diffblue commented on src/analyses/dependence_graph.h:
```diff
@@ -257,7 +257,8 @@ class dependence_grapht:
     for(const auto &location_state : state_map)
     {
-      location_state.second.populate_dep_graph(*this, location_state.first);
+      static_cast<dep_graph_domaint &>(*(location_state.second))
```
Could also use `std::static_pointer_cast`, then you won't need the extra dereference

martin-cs replied:
Done.

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -431,16 +431,25 @@ class ait:public ai_baset
+      it = p.first;
+    }
+
+    return *(it->second);
```
It's a bit unfortunate that we have to assume the pointer is non-null here; C++ GSL has a `not_null` wrapper around pointers that makes such things more explicit. Maybe we should add it to our `util` library as well?

martin-cs replied:
Maybe but I think that is beyond the scope of this PR.

Commit:
Move a function so that the ordering in .cpp matches the .h

Commit:
Thread a single history through all initialisation and fixedpoints

This is needed to support histories that carry information across
function calls.  It requires some changes to the internal API in
ai_baset but should not give any changes to the results computed.

xbauch commented on src/analyses/ai_storage.h:
```diff
@@ -0,0 +1,125 @@
+
+public:
+  std::shared_ptr<const statet> abstract_state_before(
+    locationt t,
```
Could the variable be named `l`, for consistency?

martin-cs replied:
Changed (but at a later patch so it might not show up here).

hannes-steffenhagen-diffblue replied:
Rather, could it just be called `location`?

xbauch commented on src/analyses/ai.cpp:
```diff
@@ -248,38 +248,28 @@ bool ai_baset::visit(
+    locationt to_l = std::next(l);
     DATA_INVARIANT(
-      *(goto_program.get_successors(l).begin()) == std::next(l),
+      *(goto_program.get_successors(l).begin()) == to_l,
```
Inline maybe? Seems we never use `to_l` again.

martin-cs replied:
Inlined.

xbauch commented on src/analyses/ai.cpp:
```diff
@@ -327,118 +317,174 @@ bool ai_baset::visit_edge(
+    }
+    else
+    {
+      // Function pointers are not currently supported and must be removed
```
This is a strange structure of code: do you plan on implementing this else-branch later?

martin-cs replied:
Yes.  I wanted to keep this patch set functionality neutral but with the work you and your team have done, this is *very* much on the cards.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2942e99b83024105b0fac33b507dca1d94967aef).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124408878

xbauch commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,237 @@
+#include <util/xml.h>
+
+/// A history object is an abstraction / representation of the control
+/// control-flow part of a set of traces.  The simplest example is a single
```
`control control-flow` .. just checking if this is intentional.

martin-cs replied:
Nope; editing error.  Fixed.

xbauch commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,237 @@
+      return *l < *r;
+    }
+  };
+  typedef std::set<trace_ptrt, compare_historyt> trace_sett;
```
I guess some people would prefer the `unordered` variant. With `shared_ptr` you'd get `std::hash` support for free.

martin-cs replied:
I would prefer to keep it ordered.  Histories have to have an ordering for working out which to propagate first in the abstract interpreter (see `working_sett`) and I would rather keep to just one kind of set.

hannes-steffenhagen-diffblue replied:
FWIW it's worth commenting if there was a deliberate choice to use an ordered map here, lest someone 'improve' it in the future (and it passing the regression tests by accident).

martin-cs replied:
I could write a test case that will break if the order of exploration / working_sett is changed substantially?  Would that help?

hannes-steffenhagen-diffblue replied:
@martin-cs The problem with tests like this is that if an unordered collection is used you can always get the ordering you _meant_ to get by sheer accident. This has bitten us in other places in the past. I think it's fair enough to just leave a comment here and hopefully that should at the very least flag things up during review

martin-cs replied:
I have expanded the comment to make it clearer why the ordering and per-history control of it is important.

xbauch commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,237 @@
+  ///     other than the call location) or undesireable (omitting some traces)
+  ///     POSTCONDITION(IMPLIES(result.first == BLOCKED,
+  ///                           result.second == nullptr()));
+  typedef std::pair<step_status, trace_ptrt> step_returnt;
```
Could this `typedef` precede the above comment?

martin-cs replied:
Re-ordered.

xbauch commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,237 @@
+  ///     POSTCONDITION(IMPLIES(result.first == NEW,
+  ///                           result.second.use_count() == 1 ));
+  ///  2. Merge with an existing history
+  ///     This will mean the analyser will continue of the domain is modified
```
Could you please rephrase this line?

martin-cs replied:
"This means the analyser will only continue if the domain is updated"

xbauch commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,237 @@
+  /// and then widen in a domain specific way.  However it could be used in
+  /// other ways (such as the transformer).  The key idea is that the history
+  /// tracks / should know when to widen and when to continue.
+  virtual bool widen(const ai_history_baset &other) const
```
If this is just a predicate, then maybe it could be called something like `should_widen`. Just `widen` sounds like an imperative.

martin-cs replied:
OK; `should_widen` it is.

xbauch commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,237 @@
+
+/// The common case of history is to only care about where you are now,
+/// not how you got there!
+/// Invariants are not checkable due to C++...
```
But you could make it a precondition to constructors.

martin-cs replied:
Not easily.  Non-debugging implementations of iterators don't seem to have a way of checking "is this dereferenceable".  I could pass the `goto_programt` or at least the end marker but that is kind of ugly from an interface point of view and doesn't fully check the "is dereferenceable" part as it won't catch invalidated iterators, etc.  So I think I would rather leave it as a comment and let the CI's iterator debugging builds catch if anyone gets it wrong.

martin-cs commented:
@xbauch : I realise I suggested reviewing patch by patch but, I think one or two of the things you have flagged are fixed / replaced / obsoleted by later patches.  If this is the case, do you want to me to fix the original patches or just comment that it is fixed later?

xbauch commented:
@martin-cs Comment will do, thanks.

xbauch commented on src/analyses/ai_storage.h:
```diff
@@ -103,7 +178,18 @@ class location_sensitive_storaget : public ai_storage_baset
+
+  // For backwards compatability
+  // Care should be exercised in using this.  It is possible to create domains
+  // without the any corresponding history.  This can lead to somewhat
```
-> "without the corresponding history."

martin-cs replied:
"It is possible to create domains without any corresponding history object(s)"

xbauch commented on src/analyses/ai.h:
```diff
@@ -96,6 +104,29 @@ class ai_baset
+    /// PRECONDITION(l is dereferenceable && std::next(l) is dereferenceable)
+    /// Check relies on a DATA_INVARIANT of goto_programs
+    INVARIANT(!l->is_end_function(), "No state after the last instruction");
+    return storage->abstract_traces_before(l);
```
The precondition suggests we should pass `std::next(l)` here. Just checking this is intentional.

martin-cs replied:
Nope; you are correct and that is a bug.  Fixed.  Documentation and preconditions, even not enforced ones, are useful!

xbauch commented on src/analyses/ai.h:
```diff
@@ -28,10 +61,60 @@ Author: Daniel Kroening, kroening@kroening.com
+/// the abstract interpreter.  The "tasks" that goto-analyze uses are good
+/// examples of how to do this.
+///
+/// From a development point of view, there are several directions in which
```
I wonder if parts of this documentation wouldn't be better suited for a README file.

martin-cs replied:
Maybe.  At some point I want to write a paper describing how this works in detail, the interfaces, etc.  The goal of this commit was to keep the documentation up to date and keep the "if you want to do X then you should ..." comments.  Before that we need to have it merged and working!

xbauch reviewed (approved):
Only small nitpicks. Could be split into several PRs.

martin-cs commented:
Thanks @xbauch, I have pushed an extra commit that fixes the issues you raised.  Let me know if there is more that you want.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d5ba77b00ae0933fce4ef630cf3adb9b9cadc2a6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124808826

martin-cs commented:
Apologies @xbauch , I've had to rebase and alter things because #5064 was merged so I think a number of your comments have been marked as outdated.  I folded the changes you requested in wherever possible.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 304180ed4e527b4a88d5cd8f0d8a59fbb07a3230).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124954646

hannes-steffenhagen-diffblue commented on src/analyses/ai_domain.h:
```diff
@@ -122,4 +184,74 @@ class ai_domain_baset
+  {
+  }
+
+  virtual std::unique_ptr<statet> make_domain(locationt l) const = 0;
```
Why are these `unique_ptr` if they're supposed to be used as `shared_ptr`?

martin-cs replied:
The domain factory creates them and hands over the responsibility for their deallocation to the caller.  To me a `unique_ptr` is a clearer and more precise expression of this contract.  There are places (in `ait::visit`) where these are used to create temporary domains that do not have shared pointers to them.  Also, although `ai_storage_baset::abstract_state_before` returns a shared pointer, there is no need for it to be the same domain it stores; these can be stored using `unique_ptr`'s if you want.

hannes-steffenhagen-diffblue commented on src/analyses/dependence_graph.h:
```diff
@@ -242,6 +242,10 @@ class dependence_grapht:
 
+    // The dependency graph requires that all nodes are explicitly created
+    forall_goto_program_instructions(i_it, goto_program)
+      get_state(i_it).make_bottom();
```
:pick: I know you're just moving this, but I think this is just equivalent to doing a regular for each loop over `goto_program.instructions`, might as well change that while we're at it.

martin-cs replied:
I need the iterator (i.e. the `locationt`) rather than the instruction themselves.  Is there a version of the ranged-for construct that supports something like `for(goto_programt::targett i_it : goto_program.instructions)`?

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -408,17 +409,18 @@ class ait:public ai_baset
     }
 
-    return util_make_unique<domainT>(it->second);
+    return std::make_shared<domainT>(domainT(it->second));
```
If everything is `shared_ptr` why not in the `state_map` itself?

martin-cs replied:
??? By the end of the PR it is...?  I have tried to break into smaller commits so that this is more reviewable and so that each commit has a single purpose.

hannes-steffenhagen-diffblue commented on src/analyses/dependence_graph.h:
```diff
@@ -278,17 +279,21 @@ class dependence_grapht:
+    typename state_mapt::const_iterator it = state_map.find(l);
+    if(it == state_map.end())
     {
+      auto d = std::make_shared<dep_graph_domaint>();
```
I'd have a preference for avoiding single letter variable names where reasonable names can be given.

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -433,16 +433,25 @@ class ait:public ai_baset
 protected:
   /// Map from locations to domain elements, for the results of a static
   /// analysis.
+  typedef std::shared_ptr<statet> s_ptrt;
```
`state_ptrt`?

martin-cs replied:
I think this gets fixes later commit.

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -439,14 +443,17 @@ class ait:public ai_baset
     if(it == state_map.end())
     {
-      auto p = state_map.insert(std::make_pair(l, std::make_shared<domainT>()));
+      std::shared_ptr<statet> d(domain_factory->make_domain(l));
```
`domain`, `domain_factory->make_domain(location)`?

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -439,14 +443,17 @@ class ait:public ai_baset
     {
-      auto p = state_map.insert(std::make_pair(l, std::make_shared<domainT>()));
+      std::shared_ptr<statet> d(domain_factory->make_domain(l));
+      auto p = state_map.insert(std::make_pair(l, d));
```
`insertion_result`?

hannes-steffenhagen-diffblue commented on src/analyses/invariant_propagation.cpp:
```diff
@@ -15,6 +15,39 @@ Author: Daniel Kroening, kroening@kroening.com
+  }
+
+private:
+  invariant_propagationt &ip;
```
Why is this a reference rather than a constant pointer like in the other factory?

martin-cs replied:
Ideally they should all be references but I wanted to keep the changes as minimal and obviously correct as possible.

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -36,7 +36,8 @@ class ai_baset
   typedef goto_programt::const_targett locationt;
 
-  ai_baset()
+  explicit ai_baset(std::unique_ptr<ai_domain_factory_baset> &&df)
```
Hm, so this means now it's impossible to implement an `ait` without a factory? I think this was possible before.

martin-cs replied:
It was possible before and remains possible.  `ait<domainT>` has a constructor that creates a suitable factory (using `domainT::domainT()`).  This PR should be source compatible with the previous code base.

hannes-steffenhagen-diffblue commented on src/analyses/ai.h:
```diff
@@ -406,70 +416,36 @@ class ait:public ai_baset
-    typename state_mapt::const_iterator it = state_map.find(t);
-    if(it == state_map.end())
-      return domain_factory->make_domain(t);
+    auto p(storage->abstract_state_before(l, *domain_factory));
```
Could you change this to `auto p = storage->....`? or `auto p{storage->....}`? With the combination of `auto` and the parentheses I had to do a double take here.

Also what's `p`?

martin-cs replied:
Now `auto p = storage->abstract_state_before(l, *domain_factory);`.  Better?

hannes-steffenhagen-diffblue commented on src/analyses/Makefile:
```diff
@@ -1,5 +1,6 @@
 SRC = ai.cpp \
       ai_domain.cpp \
+      ai_history.cpp \
```
Commit message on 7ba3feb41979dd9b6763cc2825a559e75641e1fb: s/compatable/compatible

Also, this seems like it could be a whole separate PR on its own

martin-cs replied:
It's really the key commit for this whole PR.  All of the other commits are just there to make it easier / simpler.  I didn't want to submit those separately as many of them are kind of pointless on their own.  There have already been maybe 4 PRs broken out of this, with more to come.

hannes-steffenhagen-diffblue reviewed (commented):
Some comments. Mostly nitpicks about spelling and naming and things like that. It's a lot of changes to make in a single PR though

martin-cs commented:
@hannes-steffenhagen-diffblue I appreciate that opinions vary on variable naming schemes.  I tend to use single letters for short term variables.  Also for things that are "small parameters" I tend to try to match the letter with the type, so we have 'a location l', 'a number n', 'an expression e', etc.  I think I have been consistent about this throughout the patch and followed the existing conventions in this part of the code.

If you insist that variable names are changed before it can be merged, then I will do it.  If not then I would prefer not to and I will try to be more considerate in future patches.

martin-cs commented:
@hannes-steffenhagen-diffblue I believe I have addressed all of your comments.  As some of them were mistakes in commit messages, I knew that it would break the tracking of comments to locations and require other rebasing so I rebased to the latest origin/develop and folded the code tweaks back into the original commits.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fe163e6a22ae33ecb3b9cfa47cfde79485a7ab5d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/125268131

peterschrammel commented on src/analyses/ai_domain.h:
```diff
@@ -122,4 +128,48 @@ class ai_domain_baset
+};
+
+// It would be great to have a single (templated) default implementation.
+// However, a default constructor is not part of the ai_domain_baset inteface
```
interface

martin-cs replied:
Thanks.  Fixed.

peterschrammel commented on src/analyses/invariant_propagation.h:
```diff
@@ -48,6 +42,8 @@ class invariant_propagationt:public
   typedef ait<invariant_set_domaint> baset;
 
 protected:
+  friend invariant_set_domain_factoryt;
```
Why that?

martin-cs replied:
Each `invariant_set_domain` needs access to a few of the fields of the invariant_propagation object.  This is a kinda wonky encapsulation violation but it would require a substantial refactor to remove this.  Instead what I have done is the `invariant_set_domain_factory` is created with a reference to the `invariant_propagationt` and creates the domain objects with references to the relevant parts of `invariant_propagationt`.  To do this it has to be a friend.

It's ugly but what it replaced was worse.

peterschrammel replied:
Ok, adding your insights above as comment in the code might help future refactoring efforts.

martin-cs replied:
Added as a comment.

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -333,16 +334,28 @@ class ai_baset
+    return domain_factory->copy_domain(s);
+  }
+
+  // abstract methods
```
This comment can be removed IMO.

martin-cs replied:
I think it is in a later patch.  (I am assuming you are reviewing per patch.)

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -333,16 +334,28 @@ class ai_baset
     const namespacet &ns);
 
-  // abstract methods
+  // Domain creation
```
`/// For domain creation` ?

martin-cs replied:
`/// For creating domain objects`

peterschrammel commented on src/analyses/ai_storage.h:
```diff
@@ -0,0 +1,125 @@
+class ai_storage_baset
+{
+public:
+  ai_storage_baset()
```
protected?

martin-cs replied:
Do you want the constructors of all abstract interfaces to be `protected`?

peterschrammel replied:
If their only purpose is to be called from a deriving class then yes.

martin-cs replied:
Done.

peterschrammel commented on src/analyses/ai_storage.h:
```diff
@@ -0,0 +1,125 @@
+
+  /// Notifies the storage that the user of the domain will not need the value
+  /// for this location.  After this has been called abstract_state_before may
+  /// return an over-approximation of the domain and get_state may give an
```
over-approximation of the value ?

martin-cs replied:
I've reworded this comment somewhat.

peterschrammel commented on src/analyses/ai.cpp:
```diff
@@ -327,118 +316,174 @@ bool ai_baset::visit_edge(
+    locationt l_begin = callee.instructions.begin();
 
-    working_sett working_set; // Redundant; fixpoint will add l_begin
+    working_sett ignore_working_set; // Redundant; fixpoint will add l_begin
```
Not sure I understand what the comment means

martin-cs replied:
Up until the last commit of the PR the fixpoint function has a design oddity; it starts the fixpoint calculation with the first element of the function.  What this means is that the entry in the working set that is generated by computing the call edge will be ignored and the same thing recreated when the recursive call to fixpoint is made.  The naming of the variable is intended to make that a little clearer.

The last patch adds a 'starting history' parameter to fixpoint so that history objects can be threaded through calls in the expected way rather than being destroyed and recreated.  So both the code and the comment should be gone by the time the PR is merged.

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -312,6 +312,26 @@ class ai_baset
     const goto_functionst &goto_functions,
     const namespacet &ns);
 
+  // function calls and return are special cases
```
Can these comments be made more doxygen-friendly (as generally more systematic class/function-level documentation would be appreciated)?

martin-cs replied:
Can I take "improve the documentation" along with the point about terminology below as actions for future PRs rather than things that need to be done in this PR?

peterschrammel commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,237 @@
+  /// Create a new history starting from a given location
+  /// This is used to start the analysis from scratch
+  /// PRECONDITION(l.is_dereferenceable());
+  explicit ai_history_baset(locationt)
```
Why is this empty definition needed here?

martin-cs replied:
To make clear that it is part of the the interface and to document it.

peterschrammel commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,237 @@
+{
+public:
+  typedef goto_programt::const_targett locationt;
+  typedef ai_history_baset tracet;
```
What does this typedef add? I find it rather confusing (besides, the term trace already being used for something different).

martin-cs replied:
Removed.

peterschrammel commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,237 @@
+class ai_history_factory_baset
+{
+public:
+  // Creates a new history from the given starting point
```
///

martin-cs replied:
Done.

peterschrammel commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,237 @@
+  virtual ai_history_baset::trace_ptrt bang(ai_history_baset::locationt) = 0;
+};
+
+// An easy factory implementation for histories that don't need parameters
```
///
etc

martin-cs replied:
Fixed.

peterschrammel commented on src/analyses/ai_storage.h:
```diff
@@ -35,20 +36,36 @@ class ai_storage_baset
-  typedef std::shared_ptr<const statet> state_ptrt;
+  typedef std::shared_ptr<statet> state_ptrt;
+  typedef std::shared_ptr<const statet> cstate_ptrt;
+  typedef ai_history_baset tracet;
```
I don't like `tracet`...

martin-cs replied:
Removed.

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -8,6 +8,39 @@ Author: Daniel Kroening, kroening@kroening.com
 /// \file
 /// Abstract Interpretation
+///
+/// This is the core of the abstract interpretation framework.
```
Cool. Great!!!

martin-cs replied:
:-)

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -8,6 +8,39 @@ Author: Daniel Kroening, kroening@kroening.com
+///     See ai_history.h for more information.
+///
+///  3. A domain factory, derived from ai_domain_factory_baset.
+///     This generates domain objects (derived from ai_domain_baset) which
```
We should decide on a terminology here and then use it consistently to avoid confusion. Domain, domain object, domain element, domain value, abstract value, ... all of these are used somewhere (all of which are legit - except 'domain').

martin-cs replied:
Can that be an action for another PR as I think that issue is wider than just the code touched by this PR?

chrisr-diffblue replied:
Seems reasonable to me to leave that for a different PR - maybe raising a github issue to track that it's something we should do?

martin-cs replied:
I will as part of my "things to do once everyone is happy with this PR".

peterschrammel commented on src/analyses/ai_storage.h:
```diff
@@ -168,12 +168,12 @@ class location_sensitive_storaget : public trace_map_storaget
 
   cstate_ptrt abstract_state_before(
-    locationt t,
+    locationt l,
```
Please squash with the commit that introduces this.

martin-cs replied:
I tried; it was a mess.  I will try again when doing a final rebase and squash of revision comments.

peterschrammel commented on src/analyses/ai.cpp:
```diff
@@ -156,22 +156,27 @@ xmlt ai_baset::output_xml(
+    return entry_state(f_it->second.body);
+
+  // It is not clear if this is even a well-structured goto_functionst object
+  return nullptr;
```
That should probably go into the interface docs

martin-cs replied:
I had hoped that the invariants / check whether a `goto_programt` is well structured would eventually give a testable documentation of this kind of thing.  I am still not clear about this.  The interface that calls the abstract interpreter on a single `goto_programt` by synthesising a `goto_functionst` for it to live in is kinda suspect but I didn't want to change the functionality in this patch so ... it is still there.

martin-cs commented:
@hannes-steffenhagen-diffblue and @peterschrammel : I have pushed a series of commits that should address your comments.  If they resolve your concerns I will:

1. Rebase on to the latest master.
2. Squash them into the relevant commits (along with the one Peter wanted squashed in).
3. Add "improve the doxygen quality of comments" to my to-do list.
4. Start an issue for sorting out the terminology.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 09f4e1a265d55b04a6a990120344b20ad28b0c63).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/127098399
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

martin-cs commented:
I'm guessing the @allredj bot failure is because I haven't rebased on to the latest develop.  I will see if it goes away after the rebase as this has passed before.

hannes-steffenhagen-diffblue reviewed (approved):
My preferences for variable names notwithstanding I believe this is probably OK now

martin-cs commented:
Thanks @hannes-steffenhagen-diffblue I will try to avoid single variable names in future PRs.

chrisr-diffblue commented on src/analyses/ai.cpp:
```diff
@@ -304,231 +308,223 @@ bool ai_baset::visit(
 {
+  // Has history taught us not to step here...
+  auto next = p->step(to_l, *(storage->abstract_traces_before(to_l)));
+  if(next.first == ai_history_baset::step_statust::BLOCKED)
```
This kind of thing would be easier to read if we had proper accessors rather than just "first"...

martin-cs replied:
At the moment I was using `std::pair` because I figured it was better to use the standard types than to replicate stuff but if you want I can create a `step_return_valuet` object?

chrisr-diffblue commented on src/analyses/ai.h:
```diff
@@ -51,8 +147,8 @@ class ai_baset
     initialize(function_id, goto_program);
-    entry_state(goto_program);
-    fixedpoint(function_id, goto_program, goto_functions, ns);
+    trace_ptrt p = entry_state(goto_program);
```
In this instance, and the couple that follow this, I'm not sure that having a temporary `p` actually adds much - I'd be tempted to just do `fixpoint(entry_state(blah), blah...` instead.

martin-cs replied:
Fair point but if you inline it then you have to wrap the function calls over multiple lines so this is actually a smaller change.

chrisr-diffblue commented on src/analyses/ai_domain.h:
```diff
@@ -122,4 +184,74 @@ class ai_domain_baset
+  typedef ai_domain_factory_baset::locationt locationt;
+  typedef ai_domain_factory_baset::trace_ptrt trace_ptrt;
+
+  std::unique_ptr<statet> copy_domain(const statet &s) const override
```
Is there any guidance on whether the 'copy' is/should be a deep copy, or whether the 'copy' might be a copy-on-write copy, or something in between? Also, why is this called `copy_domain` and not just `copy` ?

martin-cs replied:
How the factory and the domain handle copy is entirely up to them.  It is an implementation decision and this interface shouldn't constrain it.

Have renamed s/copy_domain/copy/ to s/make_domain/make/ .

chrisr-diffblue commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,244 @@
+  {
+    PRECONDITION(dynamic_cast<const ahistoricalt *>(&op) != nullptr);
+
+    return this->current_location()->location_number <
```
Is this always a valid comparison? Is there any circumstance where `op` could give a location number in a different function, say?

martin-cs replied:
It can and that is intentional.  As I understand it, location numbers are intended to be globally unique within a goto_model.  So comparing location numbers is the safe version of comparing `locationt`s.

chrisr-diffblue commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,244 @@
+
+  // Use default widening
+  // without history there is no reason to say any location is better than
+  // another to widen.
```
This comment seems a bit random to have here (though I understand why its here). Would it be more immediately clear if there was an explicit implementation here, even if that implementation just defers to the superclass implementation?

martin-cs replied:
I could but it is an inherited virtual function so that kind of is duplication.  I'll do it if you want but I'd rather leave it as just a comment.

chrisr-diffblue commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,244 @@
+{
+public:
+  /// Creates a new history from the given starting point
+  virtual ai_history_baset::trace_ptrt bang(ai_history_baset::locationt) = 0;
```
While I appreciate the humour :-) Would  'epoch' be a better/more instantly understandable name?

martin-cs replied:
Yeah... Changed.

chrisr-diffblue commented on src/analyses/ai_storage.h:
```diff
@@ -0,0 +1,298 @@
+  /// Look up the analysis state for a given history,
+  /// instantiating a new domain if required.
+  virtual statet &
+  get_state(trace_ptrt p, const ai_domain_factory_baset &fac) = 0;
```
Nit: Indentation seems a bit whacky here? Or is that clang_format being annoying?

martin-cs replied:
I just used clang-format ...

chrisr-diffblue commented on src/analyses/ai_storage.h:
```diff
@@ -0,0 +1,298 @@
+  get_state(trace_ptrt p, const ai_domain_factory_baset &fac) = 0;
+
+  /// Reset the abstract state
+  virtual void clear()
```
Is it worth making clear (hah!) what is meant by "resetting" or "clearing" the abstract state as a concept? When might this be called, for instance?

martin-cs replied:
At the moment I am not sure that it is called.  I left it in for backwards compatibility.  I am not sure that it has a use-case that wouldn't be better served by destroying the abstract interpreter object and starting again TBH.

chrisr-diffblue commented on src/analyses/ai_storage.h:
```diff
@@ -0,0 +1,298 @@
+  /// under-approximation (forcing recomputation).
+  /// If there are multiple histories that reach this location all will be
+  /// affected
+  virtual void prune(locationt l)
```
So this is like `clear()` but for only a single location?

martin-cs replied:
Sort of.  At the moment the abstract interpreter stores all of its results in memory.  That's great if you want to be able to do random access into them after the analysis is done but if you only care about, say, proving assertions, then most of the domains could be thrown away without it being a problem.  I'm planning on using this to implement a much more memory compact analyser.

The key trick is that if you prune the domain, you can return bottom when the internal interface asks for it (forcing it to be recomputed) and top when the external interface asks for it.

chrisr-diffblue commented on src/analyses/ai_storage.h:
```diff
@@ -0,0 +1,298 @@
+    trace_mapt::iterator it = trace_map.find(p->current_location());
+    if(it == trace_map.end())
+    {
+      trace_set_ptrt s(new trace_sett());
```
Is this construct the kind of thing that should just have a static method on trace_sett? e.g. `trace_sett::new_trace_set_ptr()` and then hide it's actual constructor?

martin-cs replied:
Maybe... trace_sett is a typedef around standard library objects so it would require quite a bit of code for simplifying two calls.  If it were already a custom object then, hell yeah, but as it is I am not sure it is a net saving.

chrisr-diffblue commented on src/analyses/ai_storage.h:
```diff
@@ -0,0 +1,298 @@
+  {
+    ai_storage_baset::clear();
+    trace_map.clear();
+    return;
```
Why the return?

martin-cs replied:
I tend to put explicit return statements into things with void return type so that it is clear that it is done, I haven't just got part way through and wandered off and forgotten what I was doing.  Not that that has ever happened.

chrisr-diffblue reviewed (approved):
A bunch of minor comments/queries, and I'll vote with @hannes-steffenhagen-diffblue that I don't particularly like the single character identifiers in a lot of places - but I'm not going to block this PR on any of those.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fa1c03d1c2810fd85f3d60e0a6b749c399d5f362).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/127648219

martin-cs commented:
@chrisr-diffblue : I have responded to all of your comments and patched the ones I think need patching.  Unless you object I will continue with the PR.

@peterschrammel : If you can approve the changes I made I think we are ready to do the rebase / squash / merge.

martin-cs commented:
All comments resolved; next:

1. Rebase on to the latest master.
2. Squash them into the relevant commits (along with the one Peter wanted squashed in).
3. Add "improve the doxygen quality of comments" to my to-do list.
4. Start an issue for sorting out the terminology.

smowton commented on src/analyses/reaching_definitions.cpp:
```diff
@@ -26,9 +26,34 @@ Date: February 2013
+{
+public:
+  rd_range_domain_factoryt(
+    sparse_bitvector_analysist<reaching_definitiont> *_bv_container)
```
If this is never null, should it just be a reference?

martin-cs replied:
I converted at least one case of these when I was doing this, the others I didn't convert for reasons.  IIRC these were wanting to keep the patches as small and focused as possible and it's not quite that simple; there are some cases where it isn't that straight forward.

If it bothers you I can write a separate PR for it?

smowton commented on src/analyses/ai_domain.h:
```diff
@@ -137,6 +137,11 @@ class ai_domain_factory_baset
+
+  // Not domain construction but requires knowing the precise type of statet
+  virtual bool
+  merge(statet &dest, const statet &src, locationt from, locationt to)
```
Rather than apologise for "factory" containing a non-factory method, how about "domain-manager"?

martin-cs replied:
Yes ... probably the right name for it but ... uhhh... that's going to be a massive pain to fix as it is in comments, commit messages and so on and for much of the patch set it is a sensible name.

I am going to open a discussion issue on renaming things as @peterschrammel requested, can we deal with this as part of that?

smowton commented on src/analyses/ai.cpp:
```diff
@@ -197,16 +198,18 @@ void ai_baset::finalize()
-  working_sett::iterator i=working_set.begin();
-  locationt l=i->second;
-  working_set.erase(i);
+  // STL guarantees this will be key minimal
```
If you're relying on this you should probably static-assert that `working_sett` is the type you expect

martin-cs replied:
Done.

smowton commented on src/analyses/ai.cpp:
```diff
@@ -289,27 +294,36 @@ bool ai_baset::visit(
 bool ai_baset::visit_edge(
   const irep_idt &function_id,
-  locationt l,
+  trace_ptrt p,
```
I'd appreciate `p` being named something more informative (just "trace_ptr"?) here and elsewhere. The various "l_" things at least referred to "location" rather than just "pointer" (to what?)

martin-cs replied:
Fair point.  As there is already going to be a PR for source code text changes that don't alter the functionality, can I address this as part of that?

chrisr-diffblue replied:
I'm OK with that approach, but it might be helpful to open an issue to track all the name-related cleanup that needs to be done so that things like this can be logged and then checked in the PR for text-only changes later - otherwise I fear they might get forgotten...

smowton commented on src/analyses/ai_storage.h:
```diff
@@ -211,4 +211,82 @@ class location_sensitive_storaget : public trace_map_storaget
+class history_sensitive_storaget : public trace_map_storaget
+{
+protected:
+  typedef std::map<trace_ptrt, state_ptrt, ai_history_baset::compare_historyt>
```
Do you need the ordered-ness of `map`?

martin-cs replied:
Not directly but we already have a compare function and no hash function so I figured it was better.  Also I would prefer to avoid the reallocation that comes with growing hash maps.

martin-cs commented:
If the latest rebase passes CI and no-one objects I will merge this and open a discussion issue for fixing the source text issues.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 27f27f85b2e59c7a93421b2346551799cb65be65).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136626871

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 529d7a52be4374390f0b14616501cdb2b17a2808).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/138137061
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

martin-cs commented:
( @chrisr-diffblue re-ran the test-gen build and said the one above is a Travis failure rather than an actual issue. )
